rust/lance/src/io/exec/fts.rs

```diff
@@ -198,3 +185,130 @@ impl ExecutionPlan for FtsExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(f, "FlatFts: query={}", self.query.query)
+            }
+        }
+    }
+}
+
+impl FlatFtsExec {
+    pub fn new(
+        dataset: Arc<Dataset>,
+        column_inputs: HashMap<String, (Vec<Index>, Arc<dyn ExecutionPlan>)>,
+        query: FullTextSearchQuery,
+    ) -> Self {
+        let properties = PlanProperties::new(
+            EquivalenceProperties::new(FTS_SCHEMA.clone()),
+            Partitioning::RoundRobinBatch(1),
+            ExecutionMode::Bounded,
+        );
+        Self {
+            dataset,
+            column_inputs,
+            query,
+            properties,
+        }
+    }
+}
+
+impl ExecutionPlan for FlatFtsExec {
+    fn name(&self) -> &str {
+        "FlatFtsExec"
+    }
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        FTS_SCHEMA.clone()
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
+        self.column_inputs
+            .values()
+            .map(|(_, input)| input)
+            .collect()
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    #[instrument(name = "flat_fts_exec", level = "debug", skip_all)]
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<datafusion::execution::context::TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let query = self.query.clone();
+        let ds = self.dataset.clone();
+        let column_inputs = self.column_inputs.clone();
+
+        let stream = stream::iter(column_inputs)
+            .map(move |(column, (indices, input))| {
+                let index_meta = indices[0].clone();
+                let uuid = index_meta.uuid.to_string();
+                let query = query.clone();
+                let ds = ds.clone();
+                let context = context.clone();
+
+                async move {
+                    let index = ds.open_generic_index(&column, &uuid).await?;
+                    let index =
+                        index
+                            .as_any()
+                            .downcast_ref::<InvertedIndex>()
+                            .ok_or_else(|| {
+                                DataFusionError::Execution(format!(
+                                    "Index {} is not an inverted index",
+                                    uuid,
+                                ))
+                            })?;
+
+                    let unindexed_stream = input.execute(partition, context)?;
+                    let unindexed_result_stream =
+                        flat_bm25_search_stream(unindexed_stream, column, query, index);
```

Comment:
yeah, do this because diff index may be created at diff time, say index A could be created at the time the table had 100K rows, but index B was created at the time the table had 200K rows.

single scan is still doable, but it needs to scan "union of unindexed fragments over indexes". Any documents can be large and that would be a waste, so did scan for each index (column)

Reply:
> yeah, do this because diff index may be created at diff time, say index A could be created at the time the table had 100K rows, but index B was created at the time the table had 200K rows.

Yeah that makes sense. They might cover different fragments.

**But** in the code here, they are all sharing the `unindexed_stream`. So doesn't that make your code as written incorrect too?

Reply:
this unindexed_stream is from `input` for each index(column), each one is from [here](https://github.com/lancedb/lance/pull/3036/files#diff-c25e7065f1e84cc2421b1d8b920d96038c14557fb3daadd07932f719098bc81bR1353-R1381)

Reply:
Okay I seen now. Thanks.